File: examples/middleware/timeout_example.py
"""
Timeout Middleware Example

This example demonstrates WHY and HOW to use timeout middleware for preventing
long-running requests from blocking resources.

WHY USE TIMEOUT MIDDLEWARE?
---------------------------
- Prevent hung requests: Stop operations that take too long
- Resource protection: Free up resources from slow/stuck operations
- Predictable latency: Ensure requests complete within SLA requirements
- Fault isolation: Prevent one slow operation from affecting others

WHEN TO USE:
- External API calls with unpredictable response times
- LLM requests that may hang or take too long
- Network operations prone to hanging
- Any operation with strict latency requirements

WHEN NOT TO USE:
- Known long-running operations (use higher timeout or streaming instead)
- Operations where partial results aren't acceptable
- When timeout would cause data corruption
- Real-time systems where timeout handling adds overhead

TRADE-OFFS:
- Incomplete work: Operations are cancelled mid-execution
- Resource cleanup: May need to handle cleanup of timed-out operations
- Tuning required: Finding the right timeout value can be challenging
- False positives: Legitimate slow operations may be cancelled
"""

import asyncio
import random
from agenkit.interfaces import Agent, Message
from agenkit.middleware import TimeoutConfig, TimeoutDecorator, TimeoutError


# Simulate a slow LLM API
class SlowLLMAgent(Agent):
    """Simulates an LLM that sometimes takes too long to respond."""

    def __init__(self, response_time: float = 2.0):
        self.response_time = response_time

    @property
    def name(self) -> str:
        return "slow-llm"

    @property
    def capabilities(self) -> list[str]:
        return ["text-generation"]

    async def process(self, message: Message) -> Message:
        print(f"   LLM processing (will take {self.response_time:.1f}s)...")
        await asyncio.sleep(self.response_time)
        return Message(
            role="agent",
            content=f"Generated response for: {message.content}",
            metadata={"processing_time": self.response_time}
        )


# Simulate an unpredictable agent
class UnpredictableAgent(Agent):
    """Simulates an agent with highly variable response times."""

    def __init__(self):
        self.call_count = 0

    @property
    def name(self) -> str:
        return "unpredictable-agent"

    @property
    def capabilities(self) -> list[str]:
        return ["analysis"]

    async def process(self, message: Message) -> Message:
        self.call_count += 1
        # Sometimes fast, sometimes slow
        response_time = random.choice([0.1, 0.5, 3.0, 10.0])

        print(f"   Request #{self.call_count}: will take {response_time:.1f}s")
        await asyncio.sleep(response_time)

        return Message(
            role="agent",
            content=f"Analysis complete for: {message.content}",
            metadata={"response_time": response_time}
        )


async def example_basic_timeout():
    """Example 1: Basic timeout protection."""
    print("\n=== Example 1: Basic Timeout ===")
    print("Use case: Protect against slow LLM responses\n")

    # Create a slow agent (takes 2 seconds)
    agent = SlowLLMAgent(response_time=2.0)

    # Wrap with timeout middleware (1 second limit)
    timeout_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=1.0)
    )

    message = Message(role="user", content="Explain quantum computing")

    try:
        print("Sending request with 1s timeout...")
        response = await timeout_agent.process(message)
        print(f"✅ Response received: {response.content}")
    except TimeoutError as e:
        print(f"❌ Request timed out: {e}")
        print("   The request was cancelled to prevent blocking")


async def example_successful_with_timeout():
    """Example 2: Fast requests complete successfully."""
    print("\n=== Example 2: Fast Request Success ===")
    print("Use case: Normal operations complete within timeout\n")

    # Create a fast agent (takes 0.5 seconds)
    agent = SlowLLMAgent(response_time=0.5)

    # Wrap with timeout middleware (2 second limit)
    timeout_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=2.0)
    )

    message = Message(role="user", content="Hello!")

    try:
        print("Sending request with 2s timeout...")
        response = await timeout_agent.process(message)
        print(f"✅ Response received: {response.content}")
        print(f"   Processing time: {response.metadata['processing_time']:.1f}s")
    except TimeoutError as e:
        print(f"❌ Unexpected timeout: {e}")


async def example_multiple_requests_with_metrics():
    """Example 3: Handling multiple requests and tracking metrics."""
    print("\n=== Example 3: Multiple Requests with Metrics ===")
    print("Use case: Track success/timeout rates across many requests\n")

    agent = UnpredictableAgent()

    # Set timeout at 2 seconds
    timeout_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=2.0)
    )

    message = Message(role="user", content="Analyze this data")

    # Send 5 requests
    print("Sending 5 requests with unpredictable response times...")
    for i in range(5):
        try:
            response = await timeout_agent.process(message)
            print(f"   ✅ Request {i+1}: Success")
        except TimeoutError:
            print(f"   ❌ Request {i+1}: Timeout")

    # Check metrics
    metrics = timeout_agent.metrics
    print("\n📊 Metrics Summary:")
    print(f"   Total requests: {metrics.total_requests}")
    print(f"   Successful: {metrics.successful_requests}")
    print(f"   Timed out: {metrics.timed_out_requests}")
    print(f"   Success rate: {metrics.successful_requests/metrics.total_requests*100:.1f}%")

    if metrics.min_duration:
        print(f"   Min duration: {metrics.min_duration:.3f}s")
        print(f"   Max duration: {metrics.max_duration:.3f}s")
        print(f"   Avg duration: {metrics.avg_duration:.3f}s")


async def example_fallback_on_timeout():
    """Example 4: Implement fallback when timeout occurs."""
    print("\n=== Example 4: Fallback on Timeout ===")
    print("Use case: Provide cached/default response when primary agent times out\n")

    class CachedResponseAgent(Agent):
        """Fallback agent that returns cached responses."""

        @property
        def name(self) -> str:
            return "cached-response"

        @property
        def capabilities(self) -> list[str]:
            return ["text-generation"]

        async def process(self, message: Message) -> Message:
            return Message(
                role="agent",
                content="[Cached response] I'm a general assistant. How can I help?",
                metadata={"source": "cache"}
            )

    # Primary agent (slow)
    primary = SlowLLMAgent(response_time=5.0)
    primary_with_timeout = TimeoutDecorator(
        primary,
        TimeoutConfig(timeout=1.0)
    )

    # Fallback agent (fast)
    fallback = CachedResponseAgent()

    message = Message(role="user", content="Hello!")

    print("Trying primary agent with 1s timeout...")
    try:
        response = await primary_with_timeout.process(message)
        print(f"✅ Primary response: {response.content}")
    except TimeoutError as e:
        print(f"⚠️  Primary timed out: {e}")
        print("   Falling back to cached response...")
        response = await fallback.process(message)
        print(f"✅ Fallback response: {response.content}")


async def example_streaming_timeout():
    """Example 5: Timeout with streaming responses."""
    print("\n=== Example 5: Streaming with Timeout ===")
    print("Use case: Apply timeout to entire stream duration\n")

    class StreamingAgent(Agent):
        """Agent that streams responses over time."""

        def __init__(self, chunk_delay: float = 0.5, num_chunks: int = 5):
            self.chunk_delay = chunk_delay
            self.num_chunks = num_chunks

        @property
        def name(self) -> str:
            return "streaming-agent"

        @property
        def capabilities(self) -> list[str]:
            return ["streaming", "text-generation"]

        async def process(self, message: Message) -> Message:
            return Message(role="agent", content="Full response")

        async def stream(self, message: Message):
            """Stream responses chunk by chunk."""
            for i in range(self.num_chunks):
                await asyncio.sleep(self.chunk_delay)
                yield Message(
                    role="agent",
                    content=f"Chunk {i+1}/{self.num_chunks}: Processing..."
                )

    # Create streaming agent (5 chunks × 0.5s = 2.5s total)
    agent = StreamingAgent(chunk_delay=0.5, num_chunks=5)

    # Wrap with 2-second timeout
    timeout_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=2.0)
    )

    message = Message(role="user", content="Generate long response")

    try:
        print("Streaming with 2s timeout (should timeout after ~2s)...")
        async for chunk in timeout_agent.stream(message):
            print(f"   📦 {chunk.content}")
    except TimeoutError as e:
        print(f"❌ Stream timed out: {e}")
        print("   Timeout applies to entire stream duration")


async def example_different_timeout_strategies():
    """Example 6: Different timeout strategies for different use cases."""
    print("\n=== Example 6: Timeout Strategies ===")
    print("Use case: Choose appropriate timeout based on operation type\n")

    agent = SlowLLMAgent(response_time=1.5)

    # Strategy 1: Strict timeout for interactive responses
    interactive_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=1.0)  # 1 second
    )

    # Strategy 2: Relaxed timeout for background processing
    background_agent = TimeoutDecorator(
        agent,
        TimeoutConfig(timeout=5.0)  # 5 seconds
    )

    message = Message(role="user", content="Process this")

    # Interactive request (strict timeout)
    print("Interactive request (1s timeout):")
    try:
        response = await interactive_agent.process(message)
        print(f"   ✅ Success: {response.content}")
    except TimeoutError:
        print("   ❌ Timeout: Too slow for interactive use")

    # Background request (relaxed timeout)
    print("\nBackground request (5s timeout):")
    try:
        response = await background_agent.process(message)
        print(f"   ✅ Success: {response.content}")
    except TimeoutError:
        print("   ❌ Timeout: Even background processing failed")


async def main():
    """Run all examples."""
    print("=" * 60)
    print("Timeout Middleware Examples")
    print("=" * 60)

    await example_basic_timeout()
    await example_successful_with_timeout()
    await example_multiple_requests_with_metrics()
    await example_fallback_on_timeout()
    await example_streaming_timeout()
    await example_different_timeout_strategies()

    print("\n" + "=" * 60)
    print("Examples complete!")
    print("=" * 60)
    print("\nKey Takeaways:")
    print("1. Set timeouts based on SLA requirements and operation type")
    print("2. Implement fallback strategies for timeout scenarios")
    print("3. Monitor metrics to tune timeout values")
    print("4. Use different timeouts for interactive vs background operations")
    print("5. Remember: timeout applies to entire stream duration")


if __name__ == "__main__":
    asyncio.run(main())

File: agenkit-zig/examples/middleware/timeout_example.zig
/// Example demonstrating TimeoutMiddleware usage
///
/// Shows how to:
/// 1. Configure timeout behavior
/// 2. Use method-specific timeouts
/// 3. Track timeout metrics
/// 4. Handle timeout errors

const std = @import("std");
const agenkit = @import("agenkit");

// Slow agent that takes variable time to respond
const SlowAgent = struct {
    allocator: std.mem.Allocator,
    delay_ms: u64,

    pub fn init(allocator: std.mem.Allocator, delay_ms: u64) !*SlowAgent {
        const self = try allocator.create(SlowAgent);
        self.* = SlowAgent{
            .allocator = allocator,
            .delay_ms = delay_ms,
        };
        return self;
    }

    pub fn agent(self: *SlowAgent) agenkit.Agent {
        return agenkit.Agent{
            .ptr = self,
            .vtable = &.{
                .name = nameImpl,
                .capabilities = capabilitiesImpl,
                .process = processImpl,
                .introspect = introspectImpl,
                .deinit = deinitImpl,
            },
        };
    }

    fn nameImpl(ptr: *anyopaque) []const u8 {
        _ = ptr;
        return "slow_agent";
    }

    fn capabilitiesImpl(ptr: *anyopaque, allocator: std.mem.Allocator) std.mem.Allocator.Error![]const []const u8 {
        _ = ptr;
        const caps = try allocator.alloc([]const u8, 1);
        caps[0] = "slow";
        return caps;
    }

    fn processImpl(ptr: *anyopaque, message: agenkit.Message) agenkit.AgentError!agenkit.Result {
        const self: *SlowAgent = @ptrCast(@alignCast(ptr));

        // Check if message has custom delay in metadata
        const delay_ms = if (message.metadata.get("delay_ms")) |delay_str|
            std.fmt.parseInt(u64, delay_str, 10) catch self.delay_ms
        else
            self.delay_ms;

        std.debug.print("  [Slow Agent] Processing (will take {d}ms)...\n", .{delay_ms});

        // Simulate slow processing
        std.time.sleep(delay_ms * std.time.ns_per_ms);

        std.debug.print("  [Slow Agent] Done!\n", .{});

        // Echo back the message
        return agenkit.Result{ .ok = message };
    }

    fn introspectImpl(ptr: *anyopaque, allocator: std.mem.Allocator) std.mem.Allocator.Error!agenkit.IntrospectionResult {
        _ = ptr;
        return agenkit.createDefaultIntrospectionResult(allocator, "slow_agent", &.{"slow"});
    }

    fn deinitImpl(ptr: *anyopaque) void {
        const self: *SlowAgent = @ptrCast(@alignCast(ptr));
        self.allocator.destroy(self);
    }
};

pub fn main() !void {
    var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    defer _ = gpa.deinit();
    const allocator = gpa.allocator();

    std.debug.print("\n=== Timeout Middleware Example ===\n\n", .{});

    // Example 1: Basic timeout
    std.debug.print("Example 1: Basic Timeout (1s timeout, 500ms operation)\n", .{});
    try basicTimeoutExample(allocator);

    std.debug.print("\n", .{});

    // Example 2: Timeout triggers
    std.debug.print("Example 2: Timeout Triggers (500ms timeout, 1s operation)\n", .{});
    try timeoutTriggersExample(allocator);

    std.debug.print("\n", .{});

    // Example 3: Method-specific timeouts
    std.debug.print("Example 3: Method-Specific Timeouts\n", .{});
    try methodSpecificTimeoutsExample(allocator);

    std.debug.print("\n", .{});

    // Example 4: Metrics tracking
    std.debug.print("Example 4: Metrics Tracking (multiple requests)\n", .{});
    try metricsExample(allocator);

    std.debug.print("\n=== All examples completed successfully ===\n", .{});
}

fn basicTimeoutExample(allocator: std.mem.Allocator) !void {
    // Create a slow agent (500ms delay)
    var slow = try SlowAgent.init(allocator, 500);
    defer slow.agent().deinit();

    // Wrap with timeout middleware (1000ms timeout)
    const config = agenkit.middleware.TimeoutConfig{
        .timeout_ms = 1000,
    };
    var timeout_agent = try agenkit.middleware.TimeoutDecorator.init(allocator, slow.agent(), config);
    defer timeout_agent.agent().deinit();

    // Create a test message
    var message = try agenkit.Message.withText(allocator, .user, "Test message");
    defer message.deinit();

    // Process with timeout (should succeed)
    std.debug.print("  Processing message (should succeed)...\n", .{});
    const result = try timeout_agent.agent().process(message);
    var response = try result.unwrap();
    defer response.deinit();

    const metrics = timeout_agent.metrics();
    std.debug.print("  Result: SUCCESS\n", .{});
    std.debug.print("  Duration: ~500ms (timeout: 1000ms)\n", .{});
    std.debug.print("  Metrics: {d} requests, {d} successful, {d} timed out\n", .{
        metrics.total_requests,
        metrics.successful_requests,
        metrics.timed_out_requests,
    });
}

fn timeoutTriggersExample(allocator: std.mem.Allocator) !void {
    // Create a slow agent (1000ms delay)
    var slow = try SlowAgent.init(allocator, 1000);
    defer slow.agent().deinit();

    // Wrap with timeout middleware (500ms timeout)
    const config = agenkit.middleware.TimeoutConfig{
        .timeout_ms = 500,
    };
    var timeout_agent = try agenkit.middleware.TimeoutDecorator.init(allocator, slow.agent(), config);
    defer timeout_agent.agent().deinit();

    // Create a test message
    var message = try agenkit.Message.withText(allocator, .user, "Test message");
    defer message.deinit();

    // Process with timeout (should timeout)
    std.debug.print("  Processing message (should timeout)...\n", .{});
    const result = timeout_agent.agent().process(message);

    if (result) |_| {
        std.debug.print("  ERROR: Should have timed out!\n", .{});
        return error.TestFailed;
    } else |err| {
        if (err == agenkit.AgentError.Timeout) {
            std.debug.print("  Result: TIMEOUT (as expected)\n", .{});

            const metrics = timeout_agent.metrics();
            std.debug.print("  Metrics: {d} requests, {d} timed out\n", .{
                metrics.total_requests,
                metrics.timed_out_requests,
            });
        } else {
            return err;
        }
    }
}

fn methodSpecificTimeoutsExample(allocator: std.mem.Allocator) !void {
    // Create a slow agent
    var slow = try SlowAgent.init(allocator, 100);
    defer slow.agent().deinit();

    // Configure method-specific timeouts
    const method_timeouts = [_]agenkit.middleware.MethodTimeout{
        .{ .name = "fast", .timeout_ms = 200 },
        .{ .name = "slow", .timeout_ms = 2000 },
    };

    const config = agenkit.middleware.TimeoutConfig{
        .timeout_ms = 500, // default
        .method_timeouts = &method_timeouts,
    };
    var timeout_agent = try agenkit.middleware.TimeoutDecorator.init(allocator, slow.agent(), config);
    defer timeout_agent.agent().deinit();

    // Test "fast" method (200ms timeout)
    std.debug.print("  Testing 'fast' method (200ms timeout, 100ms operation)...\n", .{});
    {
        var message = try agenkit.Message.withText(allocator, .user, "Fast operation");
        defer message.deinit();
        try message.metadata.put("method", "fast");

        const result = try timeout_agent.agent().process(message);
        var response = try result.unwrap();
        response.deinit();
        std.debug.print("    Result: SUCCESS\n", .{});
    }

    // Test "slow" method (2000ms timeout)
    std.debug.print("  Testing 'slow' method (2000ms timeout, 1500ms operation)...\n", .{});
    {
        var message = try agenkit.Message.withText(allocator, .user, "Slow operation");
        defer message.deinit();
        try message.metadata.put("method", "slow");
        try message.metadata.put("delay_ms", "1500");

        const result = try timeout_agent.agent().process(message);
        var response = try result.unwrap();
        response.deinit();
        std.debug.print("    Result: SUCCESS\n", .{});
    }

    const metrics = timeout_agent.metrics();
    std.debug.print("  Total requests: {d}, all successful: {d}\n", .{
        metrics.total_requests,
        metrics.successful_requests,
    });
}

fn metricsExample(allocator: std.mem.Allocator) !void {
    // Create a slow agent
    var slow = try SlowAgent.init(allocator, 100);
    defer slow.agent().deinit();

    // Configure timeout
    const config = agenkit.middleware.TimeoutConfig{
        .timeout_ms = 500,
    };
    var timeout_agent = try agenkit.middleware.TimeoutDecorator.init(allocator, slow.agent(), config);
    defer timeout_agent.agent().deinit();

    // Process 10 messages with varying delays
    std.debug.print("  Processing 10 messages with varying delays...\n", .{});
    const delays = [_]u64{ 100, 200, 150, 300, 400, 100, 600, 200, 250, 700 };

    var i: usize = 0;
    while (i < delays.len) : (i += 1) {
        var message = try agenkit.Message.withText(allocator, .user, "Test message");
        defer message.deinit();

        const delay_str = try std.fmt.allocPrint(allocator, "{d}", .{delays[i]});
        defer allocator.free(delay_str);
        try message.metadata.put("delay_ms", delay_str);

        const result = timeout_agent.agent().process(message) catch |err| {
            if (err == agenkit.AgentError.Timeout) {
                std.debug.print("    Message {d}: TIMEOUT (delay: {d}ms)\n", .{ i + 1, delays[i] });
                continue;
            }
            return err;
        };
        var response = try result.unwrap();
        response.deinit();
    }

    // Print final metrics
    const metrics = timeout_agent.metrics();
    std.debug.print("\n  Final Metrics:\n", .{});
    std.debug.print("    Total requests:      {d}\n", .{metrics.total_requests});
    std.debug.print("    Successful:          {d}\n", .{metrics.successful_requests});
    std.debug.print("    Timed out:           {d}\n", .{metrics.timed_out_requests});
    std.debug.print("    Failed:              {d}\n", .{metrics.failed_requests});

    if (metrics.min_duration_ms) |min| {
        std.debug.print("    Min duration:        {d}ms\n", .{min});
    }
    if (metrics.max_duration_ms) |max| {
        std.debug.print("    Max duration:        {d}ms\n", .{max});
    }
    if (metrics.avgDuration()) |avg| {
        std.debug.print("    Avg duration:        {d:.0}ms\n", .{avg});
    }
    if (metrics.timeoutRate()) |rate| {
        std.debug.print("    Timeout rate:        {d:.1}%\n", .{rate});
    }
    if (metrics.errorRate()) |err_rate| {
        std.debug.print("    Error rate:          {d:.1}%\n", .{err_rate});
    }
}
